chore: add data layer piece for references

ryanfang5 opened the pull request.

ryanfang5 commented:
I have marked a few endpoints that I believe are currently missing tests, which may have unknown validation errors.

igboyes commented on virtool/references/data.py:
```diff
@@ -0,0 +1,658 @@
+
+        return ReferenceUser(**await extend_user(self._mongo, subdocument))
+
+    async def edit_user(
```
```suggestion
    async def update_user(
```

igboyes commented on virtool/references/data.py:
```diff
@@ -0,0 +1,658 @@
+
+        return ReferenceUser(**await extend_user(self._mongo, subdocument))
+
+    async def remove_user(self, ref_id: str, user_id: str, req):
```
```suggestion
    async def delete_user(self, ref_id: str, user_id: str, req):
```

igboyes commented on virtool/references/data.py:
```diff
@@ -0,0 +1,658 @@
+
+        return IndexMinimal(**document)
+
+    async def find_groups(self, ref_id: str) -> ReferenceGroup:
```
```suggestion
    async def list_groups(self, ref_id: str) -> ReferenceGroup:
```

igboyes commented on virtool/references/data.py:
```diff
@@ -0,0 +1,658 @@
+        if not await virtool.references.db.check_right(req, ref_id, "modify"):
+            raise InsufficientRights()
+
+        targets = data.get("targets")
+
+        if targets:
+            names = [t["name"] for t in targets]
+
+            if len(names) != len(set(names)):
+                raise ResourceConflictError(
+                    "The targets field may not contain duplicate names"
+                )
```
It seems to me this would be done better in Pydantic.

igboyes reviewed (changes requested):
Please address the comments.

Once these have been resolved, I want to add API and data layer tests for all functions that modify the database. I will identify these in the next review.

ryanfang5 commented on virtool/references/utils.py:
```diff
@@ -266,12 +267,13 @@ def get_otu_schema(require_id: bool) -> dict:
 
 
-def get_owner_user(user_id: str) -> dict:
+def get_owner_user(user_id: str, created_at: Optional[datetime]) -> dict:
```
`create_document` in references/db calls `get_owner_user`, but `created_at` is a keyword argument set to None. Therefore I just typed it with Optional and assigned it to `virtool.utils.timestamp` in case it was None.

igboyes replied:
I've reviewed this and it looks like `create_document()` is always passed a value for `created_at`. It should not be optional or have a default value of `None`.

Can you verify this and make the change accordingly.

ryanfang5 commented on virtool/references/data.py:
```diff
@@ -0,0 +1,658 @@
+                user_id=user_id,
+            )
+
+        document = await attach_computed(self._mongo, document)
+
+        await self._mongo.references.insert_one(document)
+
```
This was done to include `handle` and `administrator` to the user and users fields before adding to the database as this may fix the bug where these values do not show in the find endpoint according to https://preview.virtool.ca/api/refs

ryanfang5 commented on virtool/references/data.py:
```diff
@@ -0,0 +1,658 @@
+            projection=virtool.references.db.PROJECTION,
+        )
+
+        await apply_transforms(data["documents"], [AttachUserTransform(self._mongo)])
```
You are right, I simply copied the code without taking a closer look at it. My mistake.

ryanfang5 commented:
Unmerged edit from references/db to edit_reference in references/data for testing.

Reverted change so that the reference is inserted before attach_computed is called.

ryanfang5 commented:
Added clone and remote reference tests for POST /refs. Moved tests for creating an OTU into refs and reordered existing tests.

Also, `newer` was not set in the release when creating a remote reference, so I have hard-coded it to False each time.

igboyes commented on virtool/references/data.py:
```diff
@@ -0,0 +1,685 @@
+
+        return OTUSearchResult(**data)
+
+    async def create_otus(self, ref_id: str, data: CreateOTURequest, req, user_id: str) -> OTU:
```
```suggestion
    async def create_otu(self, ref_id: str, data: CreateOTURequest, req, user_id: str) -> OTU:
```

igboyes commented on virtool/references/data.py:
```diff
@@ -0,0 +1,685 @@
+
+        return otu
+
+    async def get_history(
```
```suggestion
    async def find_history(
```

igboyes commented on virtool/references/data.py:
```diff
@@ -0,0 +1,685 @@
+
+        raise HTTPNoContent
+
+    async def edit_reference(self, ref_id: str, data: dict) -> dict:
```
Why is this separate from the `update()` method?
